scenario/constraints/ProposalConstraint.ts

```diff
@@ -1,45 +1,144 @@
-import { Constraint, Scenario, Solution, World } from '../../plugins/scenario';
+import { Constraint, Solution, World } from '../../plugins/scenario';
 import { CometContext } from '../context/CometContext';
 import { Requirements } from './Requirements';
+import { IGovernorBravo } from '../../build/types';
+import { fetchQuery } from '../utils';
+import CometActor from '../context/CometActor';
 
-function* subsets<T>(array: T[], offset = 0): Generator<T[]> {
-  while (offset < array.length) {
-    let first = array[offset++];
-    for (let subset of subsets(array, offset)) {
-      subset.push(first);
-      yield subset;
-    }
-  }
-  yield [];
-}
 
 function debug(...args: any[]) {
   console.log(`[ProposalConstraint]`, ...args);
 }
 
+// Fund voters account with eth for sending transactions
+async function fundVoters(world: World, voters: string[], actor: CometActor) {
+  const nativeTokenAmount = world.base.allocation ?? 1.0;
+  for (let voter of voters) {
+    await actor.sendEth(voter, nativeTokenAmount);
+  }
+}
+
+type PendingProposal = { proposalId: number, startBlock: number, endBlock: number };
+
+async function getAllPendingProposals(world: World, governor: string, admin: string): Promise<PendingProposal[]> {
+  const adminSigner = await world.impersonateAddress(admin);
+  const governorAsAdmin = await world.hre.ethers.getContractAt(
+    'IGovernorBravo',
+    governor,
+    adminSigner
+  ) as IGovernorBravo;
+  const votingDelay = (await governorAsAdmin.votingDelay()).toNumber();
+  const votingPeriod = (await governorAsAdmin.votingPeriod()).toNumber();
+  const block = await world.hre.ethers.provider.getBlockNumber();
+  let filter = governorAsAdmin.filters.ProposalCreated();
+  let { recentLogs } = await fetchQuery(
+    governorAsAdmin,
+    filter,
+    block - (votingDelay + votingPeriod),
+    block,
+    block
+  );
+
+  const pendingProposals: PendingProposal[] = [];
+  if (recentLogs) {
+    for (let log of recentLogs) {
+      const [proposalId, , , , , , startBlock, endBlock] = log.args;
+      const state = await governorAsAdmin.state(proposalId);
+      // Save only pending proposals
+      if (state == 0) {
+        pendingProposals.push({
+          proposalId: proposalId.toNumber(),
+          startBlock: startBlock.toNumber(),
+          endBlock: endBlock.toNumber()
+        });
+      }
+    }
+  }
+
+  return pendingProposals;
+}
+
 export class ProposalConstraint<T extends CometContext, R extends Requirements> implements Constraint<T, R> {
   async solve(requirements: R, context: T, world: World) {
     let solutions: Solution<T>[] = [];
 
-    // XXX pull all pending proposals for the associated gov (chain)
-    //  produce all the subsets and queue/execute each batch of them
-    //   to form each solution
-    let pendingProposals = [];
+    // Only run migration for mainnet scenarios
```

Comment:
Nice, I would keep an `XXX` here since this isn't meant to be a permanent limitation

Reply:
good point, resolved